matcher/match.go

```diff
@@ -0,0 +1,100 @@
+// Package matcher provides matching positioning messages to a set of filtering rules
+// and pereating messages to a topic in case of they are matched.
+package matcher
+
+import (
+	"encoding/json"
+
+	"github.com/lvl484/positioning-filter/position"
+	"github.com/lvl484/positioning-filter/repository"
+)
+
+type matcher func(position.Position, *repository.Filter) (bool, error)
+
+type matcherFilters struct {
+	filters repository.Filters
+}
+
+func (m matcherFilters) Match(pos position.Position) (bool, error) {
+	filters, err := m.filters.AllByUser(pos.UserID)
+	if err != nil {
+		return false, err
+	}
+
+	for _, filter := range filters {
+		match := matcherByType(filter.Type)
+		matched, err := match(pos, filter)
+
+		if err != nil {
+			return false, err
+		}
+
+		if matched {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func NewMatcher(filters repository.Filters) Matcher {
+	return matcherFilters{filters: filters}
+}
+
+func matcherByType(matcherType string) matcher {
+	switch matcherType {
+	case "round":
+		return matchRound
+	case "rectangular":
+		return matchRectangular
+	default:
+		return nil
+	}
+}
+
+func matchRectangular(pos position.Position, filter *repository.Filter) (bool, error) {
+	var rfilter repository.RectangularFilter
+	if err := json.Unmarshal(filter.Configuration, &rfilter); err != nil {
+		return false, err
+	}
+
+	if !(rfilter.BottomRightLatitude > pos.Latitude) ||
+		!(rfilter.TopLeftLatitude < pos.Latitude) ||
+		!(rfilter.BottomRightLongitude < pos.Longitude) ||
+		!(rfilter.TopLeftLongitude > pos.Longitude) {
+		if filter.Reversed {
+			return true, nil
+		}
+
+		return false, nil
+	}
```

Comment:
this methid always returns with true or false in this place. dous the next code have any reason?